bn_mp_set_float.c

```diff
@@ -0,0 +1,180 @@
+#include "tommath_private.h"
+#ifdef BN_MP_SET_FLOAT_C
+/* LibTomMath, multiple-precision integer library -- Tom St Denis
+ *
+ * LibTomMath is a library that provides multiple-precision
+ * integer arithmetic as well as number theoretic functionality.
+ *
+ * The library was designed directly after the MPI library by
+ * Michael Fromberger but has been written from scratch with
+ * additional optimizations in place.
+ *
+ * SPDX-License-Identifier: Unlicense
+ */
+#include <float.h>
+#if ( (defined FLT_MAX_EXP) && (FLT_RADIX == 2) )
+/* We can use a faster method if we have an IEEE compliant machine and a working stdint.h */
+#if (((defined __STDC_IEC_559__) || (defined __GCC_IEC_559)) && (defined UINT32_MAX))
+
+/* To avoid "magic numbers". Some people don't like them, even if their meaning is obvious */
+/* Bits in mantissa without the implied bit  */
+#define SIGNIFICANT (FLT_MANT_DIG - 1)
+#define BIAS (FLT_MAX_EXP - 1)
+/* Maximal unbiased exponent */
+#define MAX_UNBIASED_EXPONENT ((2*FLT_MAX_EXP) - 1)
+/* Length of a binary32 without sign bit */
+#define PAYLOAD 31
+int mp_set_float(mp_int *a, float b)
+{
+   uint64_t frac;
+   int exp, res;
+   union {
+      float   dbl;
+      uint32_t bits;
+   } cast;
+   cast.dbl = b;
+
+   exp = (int)((unsigned)(cast.bits >> SIGNIFICANT) & (unsigned)MAX_UNBIASED_EXPONENT);
+   frac = (cast.bits & ((1UL << SIGNIFICANT) - 1UL)) | (1UL << SIGNIFICANT);
+
+   if (exp == MAX_UNBIASED_EXPONENT) { /* +-inf, NaN */
+      return MP_VAL;
+   }
+   exp -= BIAS + SIGNIFICANT;
+
+   res = mp_set_long(a, frac);
+   if (res != MP_OKAY) {
+      return res;
+   }
+
+   res = (exp < 0) ? mp_div_2d(a, -exp, a, NULL) : mp_mul_2d(a, exp, a);
+   if (res != MP_OKAY) {
+      return res;
+   }
+
+   if (((cast.bits >> PAYLOAD) != 0ULL) && (mp_iszero(a) == MP_NO)) {
+      SIGN(a) = MP_NEG;
+   }
+
```

Comment:
SIGN and mp_iszero

Reply:
What, `mp_iszero`, too? Can remember that I (or somebody else, can't remember) got hit over my/their head for using `a->used == 0` and now I'm supposed to use it again? Always this rapidly paced changes \*sigh\*

Reply:
I think the current way is to use IS_ZERO :)